clippy_lints/src/methods/mod.rs

```diff
@@ -3826,6 +3867,25 @@ fn lint_filetype_is_file(cx: &LateContext<'_>, expr: &hir::Expr<'_>, args: &[hir
     span_lint_and_help(cx, FILETYPE_IS_FILE, span, &lint_msg, None, &help_msg);
 }
 
+fn lint_from_iter(cx: &LateContext<'_>, expr: &hir::Expr<'_>, args: &[hir::Expr<'_>]) {
+    let ty = cx.typeck_results().expr_ty(expr);
+    let id = get_trait_def_id(cx, &paths::FROM_ITERATOR).unwrap();
+
+    if implements_trait(cx, ty, id, &[]) {
+        // `expr` implements `FromIterator` trait
+        let iter_expr = snippet(cx, args[0].span, "..");
+        span_lint_and_sugg(
+            cx,
+            FROM_ITER_INSTEAD_OF_COLLECT,
+            expr.span,
+            "usage of `FromIterator::from_iter`",
+            "use `.collect()` instead of `::from_iter()`",
```

Comment:
This suggestion seems incorrect if the value passed to `::from_iter` does not implement `Iterator`, only `IntoIterator`.

Reply:
Oh good catch!

@pitiK3U you need to add a check that the expr_ty implements the iterator trait.

You can test this with this:

```rust
Vec::from_iter(vec![42u32]);
```

Reply:
Do you have different example of this issue?

Changing the `snippet()` to ` snippet_with_macro_callsite()` seems to fix the example given above.
But I am not sure if it fixes the issue completely.

Reply:
It shouldn't lint at all, since `Vec` doesn't implement `Iterator`, but only `IntoIterator`. So calling `Vec::from_iter(vec![42u32]);` is valid code, since `from_iter<T>` only requires `T: IntoIterator`. But changing this to `vec![42u32].collect()` doesn't work, since `collect` is a method of the `Iterator` trait and therefore not available on `Vec`.

Reply:
> Do you have different example of this issue?

```rust
Vec::from_iter([42u32].to_vec())
```